File: protocol.go
package synchro

import (
	"encoding/json"
	"time"
)

// Operation represents the type of change in a sync operation.
type Operation int

const (
	OpInsert Operation = 1
	OpUpdate Operation = 2
	OpDelete Operation = 3
)

// String returns the string representation of an Operation.
func (o Operation) String() string {
	switch o {
	case OpInsert:
		return "create"
	case OpUpdate:
		return "update"
	case OpDelete:
		return "delete"
	default:
		return "unknown"
	}
}

// ParseOperation converts a string to an Operation.
func ParseOperation(s string) (Operation, bool) {
	switch s {
	case "create":
		return OpInsert, true
	case "update":
		return OpUpdate, true
	case "delete":
		return OpDelete, true
	default:
		return 0, false
	}
}

// Record represents a single synced record.
type Record struct {
	ID        string          `json:"id"`
	TableName string          `json:"table_name"`
	BucketID  string          `json:"bucket_id,omitempty"`
	Data      json.RawMessage `json:"data"`
	UpdatedAt *time.Time      `json:"updated_at,omitempty"`
	DeletedAt *time.Time      `json:"deleted_at,omitempty"`
	Checksum  *int32          `json:"checksum,omitempty"`
}

// RegisterRequest is the request body for client registration.
type RegisterRequest struct {
	ClientID      string  `json:"client_id"`
	ClientName    *string `json:"client_name,omitempty"`
	Platform      string  `json:"platform"`
	AppVersion    string  `json:"app_version"`
	SchemaVersion int64   `json:"schema_version,omitempty"`
	SchemaHash    string  `json:"schema_hash,omitempty"`
}

// RegisterResponse is the response for client registration.
type RegisterResponse struct {
	ID                string           `json:"id"`
	ServerTime        time.Time        `json:"server_time"`
	LastSyncAt        *time.Time       `json:"last_sync_at,omitempty"`
	Checkpoint        int64            `json:"checkpoint"`
	BucketCheckpoints map[string]int64 `json:"bucket_checkpoints,omitempty"`
	SchemaVersion     int64            `json:"schema_version"`
	SchemaHash        string           `json:"schema_hash"`
}

// PullRequest is the request for pulling changes.
type PullRequest struct {
	ClientID          string           `json:"client_id"`
	Checkpoint        int64            `json:"checkpoint,omitempty"`
	BucketCheckpoints map[string]int64 `json:"bucket_checkpoints,omitempty"`
	Tables            []string         `json:"tables,omitempty"`
	Limit             int              `json:"limit,omitempty"`
	KnownBuckets      []string         `json:"known_buckets,omitempty"`
	SchemaVersion     int64            `json:"schema_version,omitempty"`
	SchemaHash        string           `json:"schema_hash,omitempty"`
}

// DefaultPullLimit is the default number of records per pull.
const DefaultPullLimit = 100

// MaxPullLimit is the maximum records per pull.
const MaxPullLimit = 1000

// PullResponse is the response for pulling changes.
type PullResponse struct {
	Changes           []Record            `json:"changes"`
	Deletes           []DeleteEntry       `json:"deletes"`
	Checkpoint        int64               `json:"checkpoint"`
	BucketCheckpoints map[string]int64    `json:"bucket_checkpoints,omitempty"`
	HasMore           bool                `json:"has_more"`
	RebuildBuckets    []string            `json:"rebuild_buckets,omitempty"`
	BucketChecksums   map[string]int32    `json:"bucket_checksums,omitempty"`
	SnapshotRequired  bool                `json:"snapshot_required,omitempty"`
	SnapshotReason    string              `json:"snapshot_reason,omitempty"`
	BucketUpdates     *BucketUpdate       `json:"bucket_updates,omitempty"`
	SchemaVersion     int64               `json:"schema_version"`
	SchemaHash        string              `json:"schema_hash"`
}

// DeleteEntry represents a deleted record in a pull response.
type DeleteEntry struct {
	ID        string `json:"id"`
	TableName string `json:"table_name"`
}

// BucketUpdate describes changes to a client's bucket subscriptions.
type BucketUpdate struct {
	Added   []string `json:"added,omitempty"`
	Removed []string `json:"removed,omitempty"`
}

// PushRequest is the request body for pushing changes.
type PushRequest struct {
	ClientID      string       `json:"client_id"`
	Changes       []PushRecord `json:"changes"`
	SchemaVersion int64        `json:"schema_version,omitempty"`
	SchemaHash    string       `json:"schema_hash,omitempty"`
}

// PushRecord represents a single record being pushed from the client.
type PushRecord struct {
	ID              string          `json:"id"`
	TableName       string          `json:"table_name"`
	Operation       string          `json:"operation"`
	Data            json.RawMessage `json:"data,omitempty"`
	ClientUpdatedAt time.Time       `json:"client_updated_at"`
	BaseUpdatedAt   *time.Time      `json:"base_updated_at,omitempty"`

	// Existing holds the current server-side row data for UPDATE/DELETE
	// operations. Populated by the engine before calling AuthorizeWriteFunc.
	// Not serialized over the wire.
	Existing map[string]any `json:"-"`
}

// PushResponse is the response for pushing changes.
type PushResponse struct {
	Accepted      []PushResult `json:"accepted"`
	Rejected      []PushResult `json:"rejected"`
	Checkpoint    int64        `json:"checkpoint"`
	ServerTime    time.Time    `json:"server_time"`
	SchemaVersion int64        `json:"schema_version"`
	SchemaHash    string       `json:"schema_hash"`
}

// PushResult represents the result of processing a single push record.
type PushResult struct {
	ID              string     `json:"id"`
	TableName       string     `json:"table_name"`
	Operation       string     `json:"operation"`
	Status          string     `json:"status"`
	ReasonCode      string     `json:"reason_code,omitempty"`
	Message         string     `json:"message,omitempty"`
	ServerVersion   *Record    `json:"server_version,omitempty"`
	ServerUpdatedAt *time.Time `json:"server_updated_at,omitempty"`
	ServerDeletedAt *time.Time `json:"server_deleted_at,omitempty"`
}

// Push status constants.
const (
	PushStatusApplied           = "applied"
	PushStatusConflict          = "conflict"
	PushStatusRejectedTerminal  = "rejected_terminal"
	PushStatusRejectedRetryable = "rejected_retryable"
)

// Pull snapshot reason constants (legacy, kept for backwards compatibility).
const (
	SnapshotReasonInitialSyncRequired = "initial_sync_required"
	SnapshotReasonCheckpointBeforeLimit = "checkpoint_before_retention"
	SnapshotReasonHistoryUnavailable    = "history_unavailable"
)

// TableMeta describes a single table's sync configuration.
type TableMeta struct {
	TableName       string   `json:"table_name"`
	PushPolicy      string   `json:"push_policy"`
	Dependencies    []string `json:"dependencies"`
	ParentTable     string   `json:"parent_table,omitempty"`
	ParentFKCol     string   `json:"parent_fk_col,omitempty"`
	UpdatedAtColumn string   `json:"updated_at_column,omitempty"`
	DeletedAtColumn string   `json:"deleted_at_column,omitempty"`
}

// TableMetaResponse is the response for the table metadata endpoint.
type TableMetaResponse struct {
	Tables        []TableMeta `json:"tables"`
	ServerTime    time.Time   `json:"server_time"`
	SchemaVersion int64       `json:"schema_version"`
	SchemaHash    string      `json:"schema_hash"`
}

// RebuildRequest is the request for rebuilding a single bucket.
type RebuildRequest struct {
	ClientID      string `json:"client_id"`
	BucketID      string `json:"bucket_id"`
	Cursor        string `json:"cursor,omitempty"`
	Limit         int    `json:"limit,omitempty"`
	SchemaVersion int64  `json:"schema_version,omitempty"`
	SchemaHash    string `json:"schema_hash,omitempty"`
}

// RebuildResponse is the response for a bucket rebuild page.
type RebuildResponse struct {
	Records        []Record `json:"records"`
	Cursor         string   `json:"cursor,omitempty"`
	Checkpoint     int64    `json:"checkpoint"`
	HasMore        bool     `json:"has_more"`
	BucketChecksum *int32   `json:"bucket_checksum,omitempty"`
	SchemaVersion  int64    `json:"schema_version"`
	SchemaHash     string   `json:"schema_hash"`
}

// DefaultRebuildLimit is the default number of records per rebuild page.
const DefaultRebuildLimit = 100

// MaxRebuildLimit is the maximum records per rebuild page.
const MaxRebuildLimit = 1000

// SnapshotRequest is the request for a full snapshot (legacy, deprecated).
type SnapshotRequest struct {
	ClientID      string          `json:"client_id"`
	Cursor        *SnapshotCursor `json:"cursor,omitempty"`
	Limit         int             `json:"limit,omitempty"`
	SchemaVersion int64           `json:"schema_version,omitempty"`
	SchemaHash    string          `json:"schema_hash,omitempty"`
}

// SnapshotCursor tracks pagination state across snapshot pages (legacy, deprecated).
type SnapshotCursor struct {
	Checkpoint int64  `json:"checkpoint"`
	TableIndex int    `json:"table_idx"`
	AfterID    string `json:"after_id"`
}

// SnapshotResponse is the response for a full snapshot page (legacy, deprecated).
type SnapshotResponse struct {
	Records       []Record        `json:"records"`
	Cursor        *SnapshotCursor `json:"cursor,omitempty"`
	Checkpoint    int64           `json:"checkpoint"`
	HasMore       bool            `json:"has_more"`
	SchemaVersion int64           `json:"schema_version"`
	SchemaHash    string          `json:"schema_hash"`
}

// Schema default kind constants.
const (
	DefaultKindNone       = "none"
	DefaultKindPortable   = "portable"
	DefaultKindServerOnly = "server_only"
)

// SchemaColumn describes a table column for client-side table creation.
type SchemaColumn struct {
	Name             string `json:"name"`
	DBType           string `json:"db_type"`
	LogicalType      string `json:"logical_type"`
	Nullable         bool   `json:"nullable"`
	DefaultSQL       string `json:"default_sql,omitempty"`
	DefaultKind      string `json:"default_kind"`
	SQLiteDefaultSQL string `json:"sqlite_default_sql,omitempty"`
	IsPrimaryKey     bool   `json:"is_primary_key"`
}

// SchemaTable describes table metadata and column definitions.
type SchemaTable struct {
	TableName       string         `json:"table_name"`
	PushPolicy      string         `json:"push_policy"`
	ParentTable     string         `json:"parent_table,omitempty"`
	ParentFKCol     string         `json:"parent_fk_col,omitempty"`
	Dependencies    []string       `json:"dependencies,omitempty"`
	UpdatedAtColumn string         `json:"updated_at_column,omitempty"`
	DeletedAtColumn string         `json:"deleted_at_column,omitempty"`
	PrimaryKey      []string       `json:"primary_key"`
	Columns         []SchemaColumn `json:"columns"`
}

// SchemaResponse returns the full sync schema contract.
type SchemaResponse struct {
	SchemaVersion int64         `json:"schema_version"`
	SchemaHash    string        `json:"schema_hash"`
	ServerTime    time.Time     `json:"server_time"`
	Tables        []SchemaTable `json:"tables"`
}

// ClientDebugResponse is the response for the debug endpoint.
type ClientDebugResponse struct {
	Client         ClientDebugInfo     `json:"client"`
	Buckets        []ServerBucketDebug `json:"buckets"`
	ChangelogStats ChangelogDebugStats `json:"changelog_stats"`
	ServerTime     time.Time           `json:"server_time"`
}

// ClientDebugInfo contains client registration state.
type ClientDebugInfo struct {
	ID              string     `json:"id"`
	ClientID        string     `json:"client_id"`
	Platform        string     `json:"platform"`
	AppVersion      string     `json:"app_version"`
	IsActive        bool       `json:"is_active"`
	LastSyncAt      *time.Time `json:"last_sync_at,omitempty"`
	LastPullAt      *time.Time `json:"last_pull_at,omitempty"`
	LastPushAt      *time.Time `json:"last_push_at,omitempty"`
	LegacyCheckpoint *int64   `json:"legacy_checkpoint,omitempty"`
	BucketSubs      []string   `json:"bucket_subs"`
}

// ServerBucketDebug contains per-bucket debug state.
type ServerBucketDebug struct {
	BucketID    string  `json:"bucket_id"`
	Checkpoint  int64   `json:"checkpoint"`
	MemberCount int     `json:"member_count"`
	Checksum    *int32  `json:"checksum,omitempty"`
}

// ChangelogDebugStats contains changelog statistics.
type ChangelogDebugStats struct {
	MinSeq       int64 `json:"min_seq"`
	MaxSeq       int64 `json:"max_seq"`
	TotalEntries int64 `json:"total_entries"`
}

File: extensions/synchro-core/src/protocol.rs
use serde::{Deserialize, Serialize};
use std::fmt;

// ---------------------------------------------------------------------------
// Operation
// ---------------------------------------------------------------------------

/// Represents the type of change in a sync operation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(i16)]
pub enum Operation {
    Insert = 1,
    Update = 2,
    Delete = 3,
}

impl Operation {
    /// Returns the wire-protocol string for this operation.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Insert => "create",
            Self::Update => "update",
            Self::Delete => "delete",
        }
    }

    /// Parses a wire-protocol string into an Operation.
    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "create" => Some(Self::Insert),
            "update" => Some(Self::Update),
            "delete" => Some(Self::Delete),
            _ => None,
        }
    }

    /// Converts from the smallint stored in the changelog.
    pub fn from_i16(v: i16) -> Option<Self> {
        match v {
            1 => Some(Self::Insert),
            2 => Some(Self::Update),
            3 => Some(Self::Delete),
            _ => None,
        }
    }

    /// Returns the smallint representation.
    pub fn to_i16(self) -> i16 {
        self as i16
    }
}

impl fmt::Display for Operation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

// ---------------------------------------------------------------------------
// Push status constants
// ---------------------------------------------------------------------------

pub const PUSH_STATUS_APPLIED: &str = "applied";
pub const PUSH_STATUS_CONFLICT: &str = "conflict";
pub const PUSH_STATUS_REJECTED_TERMINAL: &str = "rejected_terminal";
pub const PUSH_STATUS_REJECTED_RETRYABLE: &str = "rejected_retryable";

// ---------------------------------------------------------------------------
// Pull / rebuild limits
// ---------------------------------------------------------------------------

pub const DEFAULT_PULL_LIMIT: i32 = 100;
pub const MAX_PULL_LIMIT: i32 = 1000;
pub const DEFAULT_REBUILD_LIMIT: i32 = 100;
pub const MAX_REBUILD_LIMIT: i32 = 1000;

// ---------------------------------------------------------------------------
// Snapshot reason constants (legacy, kept for backwards compatibility)
// ---------------------------------------------------------------------------

pub const SNAPSHOT_REASON_INITIAL_SYNC: &str = "initial_sync_required";
pub const SNAPSHOT_REASON_CHECKPOINT_BEFORE_LIMIT: &str = "checkpoint_before_retention";
pub const SNAPSHOT_REASON_HISTORY_UNAVAILABLE: &str = "history_unavailable";

// ---------------------------------------------------------------------------
// Default kind constants (schema introspection)
// ---------------------------------------------------------------------------

pub const DEFAULT_KIND_NONE: &str = "none";
pub const DEFAULT_KIND_PORTABLE: &str = "portable";
pub const DEFAULT_KIND_SERVER_ONLY: &str = "server_only";

// ---------------------------------------------------------------------------
// Clamp helpers
// ---------------------------------------------------------------------------

/// Clamp a pull limit to [1, MAX_PULL_LIMIT], defaulting to DEFAULT_PULL_LIMIT
/// when the input is <= 0.
pub fn clamp_pull_limit(limit: i32) -> i32 {
    if limit <= 0 {
        DEFAULT_PULL_LIMIT
    } else if limit > MAX_PULL_LIMIT {
        MAX_PULL_LIMIT
    } else {
        limit
    }
}

/// Clamp a rebuild limit to [1, MAX_REBUILD_LIMIT], defaulting to
/// DEFAULT_REBUILD_LIMIT when the input is <= 0.
pub fn clamp_rebuild_limit(limit: i32) -> i32 {
    if limit <= 0 {
        DEFAULT_REBUILD_LIMIT
    } else if limit > MAX_REBUILD_LIMIT {
        MAX_REBUILD_LIMIT
    } else {
        limit
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operation_round_trip() {
        for op in [Operation::Insert, Operation::Update, Operation::Delete] {
            let s = op.as_str();
            assert_eq!(Operation::parse(s), Some(op));
            let i = op.to_i16();
            assert_eq!(Operation::from_i16(i), Some(op));
        }
    }

    #[test]
    fn operation_display() {
        assert_eq!(format!("{}", Operation::Insert), "create");
        assert_eq!(format!("{}", Operation::Update), "update");
        assert_eq!(format!("{}", Operation::Delete), "delete");
    }

    #[test]
    fn operation_parse_unknown() {
        assert_eq!(Operation::parse("unknown"), None);
        assert_eq!(Operation::from_i16(0), None);
        assert_eq!(Operation::from_i16(4), None);
    }

    #[test]
    fn clamp_pull() {
        assert_eq!(clamp_pull_limit(0), DEFAULT_PULL_LIMIT);
        assert_eq!(clamp_pull_limit(-1), DEFAULT_PULL_LIMIT);
        assert_eq!(clamp_pull_limit(50), 50);
        assert_eq!(clamp_pull_limit(2000), MAX_PULL_LIMIT);
    }

    #[test]
    fn clamp_rebuild() {
        assert_eq!(clamp_rebuild_limit(0), DEFAULT_REBUILD_LIMIT);
        assert_eq!(clamp_rebuild_limit(-1), DEFAULT_REBUILD_LIMIT);
        assert_eq!(clamp_rebuild_limit(500), 500);
        assert_eq!(clamp_rebuild_limit(5000), MAX_REBUILD_LIMIT);
    }

    #[test]
    fn from_i16_boundary_values() {
        // Negative values, i16::MIN, i16::MAX must all return None.
        assert_eq!(Operation::from_i16(-1), None);
        assert_eq!(Operation::from_i16(i16::MIN), None);
        assert_eq!(Operation::from_i16(i16::MAX), None);
        // Valid range boundaries.
        assert_eq!(Operation::from_i16(1), Some(Operation::Insert));
        assert_eq!(Operation::from_i16(3), Some(Operation::Delete));
    }

    #[test]
    fn clamp_boundary_exact() {
        // Exactly at boundaries: 1 and MAX should pass through unchanged.
        assert_eq!(clamp_pull_limit(1), 1);
        assert_eq!(clamp_pull_limit(MAX_PULL_LIMIT), MAX_PULL_LIMIT);
        assert_eq!(clamp_rebuild_limit(1), 1);
        assert_eq!(clamp_rebuild_limit(MAX_REBUILD_LIMIT), MAX_REBUILD_LIMIT);
    }
}
